### ISSUE TITLE
Full Chair Task issues

### ISSUE THREAD

**Issue by alecive** [ScazLab/rpi_integration#6]
As per https://github.com/ScazLab/rpi_integration/commit/f000197336590f662f40ceb2ba5150ee0418763d , we now have some chair task working. This is the result:

![screenshot_2017](https://user-images.githubusercontent.com/4378663/40068625-13be745e-5837-11e8-9669-725c7df976aa.png)

This are the things that I would like to see happening:

 * [x] I would like to see just two "blocks" below "build chair": "build seat" and "build back" (or something very close to it, e.g. the GET(screwdriver) can be outside of "build_back" and "build_seat" and done before everything else.
 * [x] It is kind of messy in terms of naming: what's the difference between build, fasten, and assemble? What's an "artifact" (I know what it is, but it is kind of too generic to showcase)?
 * [x] It correctly recognizes some parallel action (namely when one action is performed twice in sequence), but the "parallel" information is in the `name` field instead of the `combination` field (ie it's a sequence of actions with "parallel" in their names)
 * [x] I think that we should find a smart way to name subtakss that were not explicitly referred to by the human. E.g. "Parallelized subtasks of ASEEMBLE-TOP-OBJECT" or so on . Any ideas?
 * [x] The output json file is not indented any more!
 * [x] I would move the "HUMAN" or "ROBOT" description into the `attributes` field (that's what the `attributes` field is about, ie encapsulating corollary information about the subtask or the action that goes beyond mere naming).

@drjesseenglish  the input sequence that generated this output file is stored in `tests/in/full_chair.json` whereas the output file is in `tests/out/full_chair.json`

**Comment by drjesseenglish** [comment]
A few responses to start with:

Regarding the parallel combination field - sorry, perhaps I'm misunderstanding where that information in the HTN is supposed to be.  If a subtask has children that can be run in parallel, I thought that the children were supposed to be marked that way, and the subtask itself should be marked however is appropriate relative to its siblings.

So the "Parallel subtasks of ..." in the example above is listed as Sequential as it must be run before its siblings; but its *children* are marked as Parallel, as they can be run in either order relative to each other.  Is that not how we are supposed to mark it?

Regarding naming of unmentioned subtasks: This is actually a pretty large problem in our field of research.  We probably shouldn't impose the weight of that problem on the output here (which is why I went with the generic names that are constructed in the current output).  Ideally, what would you want that to look like?

**Comment by alecive** [comment]
> So the "Parallel subtasks of ..." in the example above is listed as Sequential as it must be run before its siblings; but its children are marked as Parallel, as they can be run in either order relative to each other. Is that not how we are supposed to mark it?

I would actually mark the parent (and only the parent) as "Parallel". That is what the `combination` field is for, i.e. "my children can be combined in any order" or whatever.

> Regarding naming of unmentioned subtasks: This is actually a pretty large problem in our field of research.

I perfectly understand this. I think for me it's more a visualization issue rather than anything else. That's why I was polling for ideas: maybe we should be having an empty label or a placeholder that is the same for every un-labeled subtask? @JakeBrawer any ideas? Maybe I can ask scaz, but that's admittedly pretty minor anyway.

Any comment on my other requests?

**Comment by alecive** [comment]
> sequential I guess? In any case, changing this is not a big deal.

you can leave it empty, or add `leaf`

> I'll look into moving HUMAN and ROBOT to the attributes field. That is probably not a big deal.

Great!

>  I would be comfortable with reducing all instances of ASSEMBLE to BUILD as part of the format of your output, if you'd prefer (this will just be part of the output, and not part of the reasoning, which might lead to some results that appear to work differently than they do).

That's fine. In my view, one "builds" a chair, "assembles" parts of the chair (e.g. assemble back), and "fastens" screws.

> Regarding the shape of the learned tree - at this point I don't think there is going to be any change happening there. Unfortunately, I have other primary work that I need to focus on - changing the shape of the tree would involve delving into the heuristics pretty significantly (unlike the other requests which are mostly formatting of output). Since it is related, we should further discuss the learning of parallelization (merged HTNs) that we talked about earlier, now that I've had time to think on it. If you have time tomorrow morning, that would be good.

Let's talk about this tomorrow morning. In short, having a meaningful HTN has higher priority than the parallelization stuff though. Because if we don't have a nicer HTN, it is going to be very difficult to explain how a human can operate on it.
Again, I don't care much about it personally, but all of this work is about making things right for people who do care :-)

**Comment by drjesseenglish** [comment]
Pushed leia/agents:latest.yale with the latest changes.

**Comment by alecive** [comment]
I'd say we are good to go with this one! Thanks @drjesseenglish
